src/slice-chop.c

```diff
@@ -380,6 +380,244 @@ static SEXP chop_fallback_shaped(SEXP x, SEXP indices, struct vctrs_chop_info in
+  // having a `NULL` common type
+  if (!null_indices) {
+    if (x_size != vec_size(indices)) {
+      Rf_errorcall(R_NilValue, "`x` and `indices` must be lists of the same size");
+    }
+    if (!vec_is_list(indices)) {
+      Rf_errorcall(R_NilValue, "`indices` must be a list of integers, or `NULL`");
+    }
+  }
+
+  ptype = PROTECT(vctrs_type_common_impl(x, ptype));
+
+  if (ptype == R_NilValue) {
+    UNPROTECT(1);
+    return R_NilValue;
+  }
+
+  x = PROTECT(vec_cast_common(x, ptype));
+
+  R_len_t out_size = 0;
+
+  SEXP x_sizes = vctrs_shared_empty_int;
+  if (null_indices) {
+    x_sizes = Rf_allocVector(INTSXP, x_size);
+  }
+  PROTECT(x_sizes);
+  int* p_x_sizes = INTEGER(x_sizes);
+
+  // `out_size` is computed from `indices` unless it is `NULL`
+  if (null_indices) {
+    for (R_len_t i = 0; i < x_size; ++i) {
+      R_len_t x_elt_size = vec_size(VECTOR_ELT(x, i));
+
+      out_size += x_elt_size;
+      p_x_sizes[i] = x_elt_size;
+    }
+  } else {
+    for (R_len_t i = 0; i < x_size; ++i) {
+      SEXP elt = VECTOR_ELT(x, i);
+
+      if (elt == R_NilValue) {
+        continue;
+      }
+
+      R_len_t index_size = vec_size(VECTOR_ELT(indices, i));
+      out_size += index_size;
+
+      // Each element of `x` is recycled to its corresponding index's size
+      elt = vec_recycle(elt, index_size, args_empty);
+      SET_VECTOR_ELT(x, i, elt);
+    }
+  }
+
+  indices = PROTECT(vec_as_indices(indices, out_size, R_NilValue));
+
+  PROTECT_INDEX out_pi;
+  SEXP out = vec_init(ptype, out_size);
+  PROTECT_WITH_INDEX(out, &out_pi);
+  out = vec_proxy(out);
+  REPROTECT(out, out_pi);
+
+  bool is_shaped = has_dim(ptype);
+  bool has_inner_names = !is_data_frame(ptype) && list_has_inner_names(x, x_size);
+
+  SEXP out_names = vctrs_shared_empty_chr;
+  if (has_inner_names) {
+    out_names = Rf_allocVector(STRSXP, out_size);
+  }
+  PROTECT(out_names);
+
+  if (null_indices) {
+    out = vec_unchop_sequentially(out, &out_names, out_pi, x, x_size, x_sizes, is_shaped, has_inner_names);
+  } else {
+    out = vec_unchop_indices(out, &out_names, out_pi, x, x_size, indices, is_shaped, has_inner_names);
+  }
+  REPROTECT(out, out_pi);
+
+  SEXP out_size_sexp = PROTECT(r_int(out_size));
+
+  out = vec_restore(out, ptype, out_size_sexp);
+  REPROTECT(out, out_pi);
+
+  if (has_inner_names) {
+    out_names = PROTECT(vec_as_names(out_names, name_repair));
+    out = vec_set_names(out, out_names);
+    REPROTECT(out, out_pi);
+    UNPROTECT(1);
+  }
+
+  UNPROTECT(7);
+  return out;
+}
+
+static SEXP vec_unchop_indices(SEXP out,
+                               SEXP* p_out_names,
+                               PROTECT_INDEX out_pi,
+                               SEXP x,
+                               R_len_t x_size,
+                               SEXP indices,
+                               bool is_shaped,
+                               bool has_inner_names) {
+  // Modified in place
+  SEXP out_names = *p_out_names;
```

Comment:
I guess you used `&` as a call site indication this is going to get mutated by vec-assign? I'm not a fan of ref and deref in this way, and that it needs a comment for clarity.

I also don't like that with #842 there is a less strong guarantee this will be mutated in place.

Maybe make `out_names` a return parameter? We'd take a pointer and update the pointer after modification. It needs to be reprotected at the call site. Alternatively just assign it as names of `out` and carry it along in attributes.

Reply:
I've refactored to make this much clearer. We no longer need the names pointer, names handling is done completely within either `vec_unchop_indices()` or `vec_unchop_sequentially()`.

I've also ensured that we are protecting and catching the output of all `vec_assign_impl()` calls